From b147195f73ca093065c1f78a42d5b9ade0930dd0 Mon Sep 17 00:00:00 2001
From: Zheng Zucheng <zhengzucheng@huawei.com>
Date: Tue, 31 May 2022 10:57:27 +0800
Subject: [PATCH 2187/2592] [Huawei] sched/cputime: Fix kabi for kernel_cpustat

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 186871

-------------------------------

CPUTIME_SOFTIRQ_IDLE/CPUTIME_IRQ_IDLE is added to cpu_usage_stat broken
the KABI of kernel_stat, so remove them and add new kernel_irqstat to
record softirq/irq time in the idle state.

Fixes: 5ef292e4be8e "([Backport] sched/cputime: account irq cputime for accurately)"
Signed-off-by: Zheng Zucheng <zhengzucheng@huawei.com>
---
 include/linux/kernel_stat.h |  2 --
 kernel/sched/cputime.c      | 22 ++++++++++++++++++----
 2 files changed, 18 insertions(+), 6 deletions(-)

diff --git a/include/linux/kernel_stat.h b/include/linux/kernel_stat.h
index 471b20fac4a6..69a9de59032b 100644
--- a/include/linux/kernel_stat.h
+++ b/include/linux/kernel_stat.h
@@ -28,8 +28,6 @@ enum cpu_usage_stat {
 	CPUTIME_STEAL,
 	CPUTIME_GUEST,
 	CPUTIME_GUEST_NICE,
-	CPUTIME_SOFTIRQ_IDLE,
-	CPUTIME_IRQ_IDLE,
 	NR_STATS,
 };
 
diff --git a/kernel/sched/cputime.c b/kernel/sched/cputime.c
index 4848c3cdaff6..2c8b512abb30 100644
--- a/kernel/sched/cputime.c
+++ b/kernel/sched/cputime.c
@@ -6,6 +6,20 @@
 
 #ifdef CONFIG_IRQ_TIME_ACCOUNTING
 
+enum irq_idle_stat {
+	CPUTIME_SOFTIRQ_IDLE,
+	CPUTIME_IRQ_IDLE,
+	NR_IRQ_STATS,
+};
+
+struct kernel_irqstat {
+	u64 irqstat[NR_IRQ_STATS];
+};
+
+DEFINE_PER_CPU(struct kernel_irqstat, kernel_irqstat);
+#define kirqstat_this_cpu this_cpu_ptr(&kernel_irqstat)
+#define kirqstat_cpu(cpu) per_cpu(kernel_irqstat, cpu)
+
 /*
  * There are no locks covering percpu hardirq/softirq time.
  * They are only modified in vtime_account, on corresponding CPU
@@ -70,11 +84,11 @@ void irqtime_account_irq(struct task_struct *curr)
 	if (hardirq_count()) {
 		irqtime_account_delta(irqtime, delta, CPUTIME_IRQ);
 		if (curr == rq->idle)
-			kcpustat_this_cpu->cpustat[CPUTIME_IRQ_IDLE] += delta;
+			kirqstat_this_cpu->irqstat[CPUTIME_IRQ_IDLE] += delta;
 	} else if (in_serving_softirq() && curr != this_cpu_ksoftirqd()) {
 		irqtime_account_delta(irqtime, delta, CPUTIME_SOFTIRQ);
 		if (curr == rq->idle)
-			kcpustat_this_cpu->cpustat[CPUTIME_SOFTIRQ_IDLE] += delta;
+			kirqstat_this_cpu->irqstat[CPUTIME_SOFTIRQ_IDLE] += delta;
 	}
 
 }
@@ -1167,8 +1181,8 @@ unsigned long long sched_get_idle_time(int cpu)
 
 #ifdef CONFIG_IRQ_TIME_ACCOUNTING
 	if (sched_clock_irqtime) {
-		hi = kcpustat_cpu(cpu).cpustat[CPUTIME_IRQ_IDLE];
-		si = kcpustat_cpu(cpu).cpustat[CPUTIME_SOFTIRQ_IDLE];
+		hi = kirqstat_cpu(cpu).irqstat[CPUTIME_IRQ_IDLE];
+		si = kirqstat_cpu(cpu).irqstat[CPUTIME_SOFTIRQ_IDLE];
 	}
 #endif /* CONFIG_IRQ_TIME_ACCOUNTING */
 
-- 
2.17.1

